### PR TITLE
Add an HttpClient interface and NodeHttpClient implementation.

### PR DESCRIPTION
**Pull request by dcr-stripe**
### Notify

r? @richardm-stripe + @tjfontaine-stripe (for extra 👀 )

### Summary

Adds an `HttpClient` interface and moves out the Node `http`/`https` request logic into a `NodeHttpClient` implementation. Update Stripe config to allow specifying a custom `HttpClient`.

This is the first step in allowing a custom client (say one that uses fetch for https://github.com/stripe/stripe-node/issues/997).

The interface is summarized in the typescript. In brief:
- `makeRequest` returns a `Response` as a promise.
- `Response#toStream(streamCompleteCallback)` returns a streamable version of the content. The Node behavior is unchanged here of just returning the raw response itself with the headers tacked on.
- `Response#toJSON()` consumes the response and returns the JSON representation as a promise.

This should effectively be a no-op, but this should be released as its own version in case there are issues.

You can see an example rough implementation of this for fetch in https://github.com/stripe/stripe-node/pull/1208.

### Test plan

Added unit tests both to `StripeResource` and to `NodeHttpClient` itself.

### CODE REVIEW

File: types/net/net.d.ts
Comment (jt-stripe) [inline types/net/net.d.ts:33]: Do we have a union type that's, like, all possible stripe resource types?

File: types/net/net.d.ts
Comment (jt-stripe) [inline types/net/net.d.ts:67]: I'm not a Typescript expert, but is this declaration required when it's explicitly saying that it implements the `HttpClient` interface?

Also is it important that the headers are a `{[key: string]: string}` but an `object` in the interface?

File: lib/net/NodeHttpClient.js
Comment (jt-stripe) [inline lib/net/NodeHttpClient.js:9]: Is there a reason not to move this into the constructor? It'd be nice to not have side effects on import in the case that this is being imported in a non-node environment.

File: lib/stripe.js
Comment (jt-stripe) [inline lib/stripe.js:54]: This will fail in non-Node environments. Maybe instead of exposing an instantiated `NodeHttpClient` we could expose a `createNodeHttpClient` function?

```js
Stripe.createNodeHttpClient = (agent) => {
   const NodeHttpClient = require('./net/NodeHttpClient');
   return new NodeHttpClient(agent);
};
```

File: lib/stripe.js
Comment (jt-stripe) [inline lib/stripe.js:102]: Then taking my suggestion above this would be 
```js
httpClient: props.httpClient || Stripe.createNodeHttpClient(agent)
```

**Review by jt-stripe — COMMENTED** [review]
A few questions about the TS definitions and a suggestion for how to expose this... I didn't carefully review all the implementation details yet!

File: types/net/net.d.ts
Comment (dcr-stripe) [reply types/net/net.d.ts]: > I'm not a Typescript expert, but is this declaration required when it's explicitly saying that it implements the HttpClient interface?

Unfortunately yes. See [this playground](https://www.typescriptlang.org/play?ssl=5&ssc=1&pln=5&pc=24#code/JYOwLgpgTgZghgYwgAgILIN4ChnICYD2AygQLYRgAWoA5gBQAeAXMiAK6kBG0AlMiwDcCwPAG4sAXyxYEAGzgBnBcgBCyYKQAOsiOXDL0GCUA) as an example. If we remove it, we get the error `Class 'NodeHttpClient' incorrectly implements interface 'HttpClient'.
  Property 'makeRequest' is missing in type 'NodeHttpClient' but required in type 'HttpClient'.

> Also is it important that the headers are a {[key: string]: string} but an object in the interface?

Good catch, that was a stale type. Both should be `object`.
`

File: types/net/net.d.ts
Comment (dcr-stripe) [reply types/net/net.d.ts:33]: I don't think so unfortunately. @richardm-stripe to confirm?

Comment (richardm-stripe) [reply types/net/net.d.ts:33]: We don't but could easily have one if we wanted.

Though `object` feels better to me here, since `HTTPClient` should be mostly stripe-domain-agnostic.

File: lib/stripe.js
Comment (dcr-stripe) [reply lib/stripe.js]: Done. Good call.

File: lib/net/NodeHttpClient.js
Comment (richardm-stripe) [reply lib/net/NodeHttpClient.js:9]: We do need a singleton -- if the user instantiates multiple stripe clients (without overriding the agent) then they should all share the same connection pool. We could lazily instantiate it when the constructor is called, though.

File: lib/net/NodeHttpClient.js
Comment (richardm-stripe) [inline lib/net/NodeHttpClient.js:122]: very nit: WDYT of 
```
module.exports = {NodeHttpClient, NodeHttpResponse}
```
putting one "on" the other feels strange to me (although I guess we put all this stuff on `Stripe`, so...)

File: lib/net/NodeHttpClient.js
Comment (jt-stripe) [reply lib/net/NodeHttpClient.js:9]: I think this is resolved by @dcr-stripe 's latest changes

File: lib/net/NodeHttpClient.js
Comment (dcr-stripe) [reply lib/net/NodeHttpClient.js]: Done.

File: lib/net/HttpClient.js
Comment (richardm-stripe) [inline lib/net/HttpClient.js:24]: This is great

File: types/net/net.d.ts
Comment (richardm-stripe) [inline types/net/net.d.ts:43]: Should we pursue generic type parameters?
```
export interface HttpClient<Response> {
  makeRequest(...): Promise<Response>
}

interface HttpClientResponse<RawResponse, Stream> {
  ...
  getRawResponse(): RawResponse,
  toStream(...): Stream,
  ...
}
```

File: test/StripeResource.spec.js
Comment (richardm-stripe) [inline test/StripeResource.spec.js:698]: I think this is missing a `.catch` so if stripe.charges.create fails the error won't be properly propagated to the test runner

File: test/StripeResource.spec.js
Comment (richardm-stripe) [inline test/StripeResource.spec.js:219]: Would there be anyway to add a test for when there's a connection error *after* the headers have been transmitted, but halfway through when the body is being transmitted?

(this should be pretty similar to the test case above on line #200).

That's the one case I can think of where (correct me if I'm wrong) I don't think we have any test coverage but it still would be good to know that this PR is consistent with the previous behavior (and that `fetch` and `nodehttpclient` will behave consistently, too)

**Review by richardm-stripe — COMMENTED** [review]
This is absolutely wonderful. In particular I think the decision to use a promise interface for HttpClient vs. more conservatively attempting to preserve the callbacky interface represents a huge improvement to the understandability of the code and will make a far better experience for implementers of this interface.

I left a few comments, though just some ideas, nothing blocking. I'll re-assign you to the PR, feel free to assign me again once you've had a look/responded and I will approve.

File: test/StripeResource.spec.js
Comment (dcr-stripe) [reply test/StripeResource.spec.js:220]: Done.

File: types/net/net.d.ts
Comment (dcr-stripe) [reply types/net/net.d.ts]: Done.

File: test/StripeResource.spec.js
Comment (dcr-stripe) [reply test/StripeResource.spec.js]: Done.

**Review by dcr-stripe — COMMENTED** [review]
Great ideas all around. PTAL @richardm-stripe

File: test/net/NodeHttpClient.spec.js
Comment (richardm-stripe) [inline test/net/NodeHttpClient.spec.js:26]:
```suggestion
    if (this._index === this._values.length) {
```

**Review by richardm-stripe — APPROVED** [review]
LGTM, (pending fixing that TS complaint about `any`)

**Pull request merged**
